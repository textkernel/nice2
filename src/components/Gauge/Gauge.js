import * as React from 'react';
import PropTypes from 'prop-types';
import { bem } from '../../utils';
import { ContentPlaceholder } from '../ContentPlaceholder';
import { CONTEXTS } from '../../constants';
import styles from './Gauge.scss';

const { block, elem } = bem('Gauge', styles);

export const Gauge = (props) => {
    const { children, context, isLoading, percentage, value, metric, ...rest } = props;

    const progress = isLoading ? 0 : percentage / 100;
    const radius = 100;
    const circumference = 2 * Math.PI * radius;
    const circumferenceHalf = circumference / 2;
    const strokeDasharrayBackground = `${circumferenceHalf} ${circumference}`;
    const strokeDasharrayForeground = `${progress * circumferenceHalf} ${circumference}`;

    return (
        <div {...rest} {...block(props)}>
            <svg {...elem('svg', props)} width="100%">
                <linearGradient id={`Gauge__gradient--${context}`}>
                    <stop className={`Gauge__gradientStart--${context}`} offset="50%" />
                    <stop className={`Gauge__gradientEnd--${context}`} offset="100%" />
                </linearGradient>
                <circle
                    {...elem('circleBackground', props)}
                    r={radius}
                    cx="50%"
                    cy="25%"
                    style={{
                        strokeDasharray: strokeDasharrayBackground,
                    }}
                />
                <circle
                    {...elem('circleForeground', props)}
                    r={radius}
                    cx="50%"
                    cy="25%"
                    style={{
                        strokeDasharray: strokeDasharrayForeground,
                    }}
                    stroke={`url(#Gauge__gradient--${context})`}
                />
            </svg>
            <div {...elem('valueWrapper', props)}>
                {isLoading ? (
                    <ContentPlaceholder height={28} {...elem('contentPlaceholder', props)} />
                ) : (
                    <span {...elem('value', props)}>
                        {value}
                        {!!metric && <span {...elem('metric', props)}>{metric}</span>}
                    </span>
                )}
            </div>
            {!!children && <div {...elem('bottom', props)}>{children}</div>}
        </div>
    );
};

Gauge.displayName = 'Gauge';

Gauge.propTypes = {
    context: PropTypes.oneOf(CONTEXTS),
    children: PropTypes.node,
    isLoading: PropTypes.bool,
    percentage: PropTypes.number,
    value: PropTypes.node,
    metric: PropTypes.node,
};

Gauge.defaultProps = {
    context: 'brand',
    children: null,
    isLoading: false,
    percentage: null,
    value: null,
    metric: null,
};
