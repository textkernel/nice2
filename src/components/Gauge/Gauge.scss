@import '../../themes/oneui/oneui.scss';

.Gauge {
    position: relative;

    &__svg {
        transform: rotate(180deg);
    }

    &__circleBackground {
        stroke: var(--color-neutral-25);
        stroke-width: var(--spacing-2x);
        stroke-linecap: round;
        fill: none;
    }

    &__circleForeground {
        stroke-width: var(--spacing-4x);
        stroke-linecap: round;
        fill: none;
        transition: all var(--transition-duration) ease-out;

        &--isProgressLoading {
            opacity: 0;
        }
    }

    &__contentPlaceholder {
        margin-bottom: 0;
    }

    &__contentWrapper {
        position: absolute;
        top: 50%;
        left: 50%;
        margin-top: var(--spacing-normal);
        transform: translate(-50%, -100%);
        text-align: center;

        &--isValueLoading {
            width: 100px;
        }
    }

    &__value {
        font-size: var(--h1-font-size);
        font-weight: var(--font-weight-bold);
    }

    &__metric {
        display: block;
        font-size: var(--font-size-normal);
        font-weight: var(--font-weight-normal);
    }

    &__bottom {
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate(-50%, 100%);
        color: var(--color-muted);
        text-transform: uppercase;
        font-weight: var(--font-weight-bold);
        font-size: var(--font-size-small);
    }

    @each $context in ($contexts) {
        &__gradientStart--#{ $context } {
            stop-color: var(--color-#{$context});
        }

        &__gradientEnd--#{ $context } {
            stop-color: var(--color-#{$context}-25);
        }
    }
}
