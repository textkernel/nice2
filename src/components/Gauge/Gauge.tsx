import * as React from 'react';
import { bem } from '../../utils';
import { ContentPlaceholder } from '../ContentPlaceholder';
import { Context, GAUGE_RADIUS } from '../../constants';
import styles from './Gauge.scss';

const { block, elem } = bem('Gauge', styles);

interface Props {
    context?: Context;
    children?: ReactNode;
    isProgressLoading?: boolean;
    isContentLoading?: boolean;
    percentage: number;
    value: NotEmptyReactNode;
    metric?: ReactNode;
}

export const Gauge: React.FC<Props> = (props) => {
    const {
        children,
        percentage,
        context,
        value,
        metric,
        isProgressLoading,
        isContentLoading,
        ...rest
    } = props;

    const percentageAdjusted = Math.max(0, Math.min(100, percentage)) / 100;
    const progress = isProgressLoading ? 0 : percentageAdjusted;
    const circumference = 2 * Math.PI * GAUGE_RADIUS;
    const circumferenceHalf = circumference / 2;

    return (
        <div {...rest} {...block(props)}>
            <svg {...elem('svg', props)} width="100%">
                <linearGradient id={`Gauge__gradient--${context}`}>
                    <stop className={`Gauge__gradientStart--${context}`} offset="50%" />
                    <stop className={`Gauge__gradientEnd--${context}`} offset="100%" />
                </linearGradient>
                <circle
                    {...elem('circleBackground', props)}
                    r={GAUGE_RADIUS}
                    cx="50%"
                    cy="25%"
                    style={{ strokeDasharray: `${circumferenceHalf} ${circumference}` }}
                />
                <circle
                    {...elem('circleForeground', props)}
                    r={GAUGE_RADIUS}
                    cx="50%"
                    cy="25%"
                    style={{ strokeDasharray: `${progress * circumferenceHalf} ${circumference}` }}
                    stroke={`url(#Gauge__gradient--${context})`}
                />
            </svg>
            <div {...elem('contentWrapper', props)}>
                {isContentLoading ? (
                    <ContentPlaceholder height={28} {...elem('contentPlaceholder', props)} />
                ) : (
                    <span {...elem('value', props)}>
                        {value}
                        {metric ? <span {...elem('metric', props)}>{metric}</span> : null}
                    </span>
                )}
            </div>
            {children ? <div {...elem('bottom', props)}>{children}</div> : null}
        </div>
    );
};

Gauge.displayName = 'Gauge';

Gauge.defaultProps = {
    children: null,
    context: 'brand',
    isProgressLoading: false,
    isContentLoading: false,
    metric: null,
};
