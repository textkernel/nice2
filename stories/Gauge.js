/* eslint-disable react/no-children-prop */
import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, number, text, select, withKnobs } from '@storybook/addon-knobs';
import { Gauge } from '@textkernel/oneui';
import { CONTEXTS } from '@textkernel/oneui/constants';

storiesOf('Molecules|Gauge', module)
    .addDecorator(withKnobs)
    .add('Gauge', () => {
        const isLoading = boolean('isLoading', false);

        return (
            <div style={{ display: 'flex' }}>
                <div style={{ width: '33%' }}>
                    <Gauge
                        context={select('Context', CONTEXTS, CONTEXTS[1])}
                        percentage={number('Percentage', 33)}
                        value={text('Value', '3')}
                        metric={text('Metric', 'Postings per job')}
                        isLoading={isLoading}
                    >
                        {isLoading ? '—' : 'Low'}
                    </Gauge>
                </div>
                <div style={{ width: '33%' }}>
                    <Gauge percentage={46} value="12" metric="Days" isLoading={isLoading}>
                        {isLoading ? '—' : 'Normal'}
                    </Gauge>
                </div>
                <div style={{ width: '33%' }}>
                    <Gauge context="primary" percentage={80} value="$54,321" isLoading={isLoading}>
                        {isLoading ? '—' : 'High'}
                    </Gauge>
                </div>
            </div>
        );
    });
