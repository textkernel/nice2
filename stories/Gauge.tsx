/* eslint-disable react/no-children-prop */
import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, number, text, select, withKnobs } from '@storybook/addon-knobs';
import { Gauge } from '@textkernel/oneui';
import { CONTEXTS } from '@textkernel/oneui/constants';

storiesOf('Molecules|Gauge', module)
    .addDecorator(withKnobs)
    .add('Gauge', () => {
        const isProgressLoading = boolean('isProgressLoading', false);
        const isContentLoading = boolean('isContentLoading', false);

        return (
            <div style={{ display: 'flex' }}>
                <div style={{ width: '33%' }}>
                    <Gauge
                        context={select('Context', CONTEXTS, CONTEXTS[1])}
                        percentage={number('Percentage', 33)}
                        value={text('Value', '3')}
                        metric={text('Metric', 'Postings per job')}
                        isProgressLoading={isProgressLoading}
                        isContentLoading={isContentLoading}
                    >
                        {isContentLoading ? '—' : 'Low'}
                    </Gauge>
                </div>
                <div style={{ width: '33%' }}>
                    <Gauge
                        percentage={46}
                        value="12"
                        metric="Days"
                        isProgressLoading={isProgressLoading}
                        isContentLoading={isContentLoading}
                    >
                        {isContentLoading ? '—' : 'Normal'}
                    </Gauge>
                </div>
                <div style={{ width: '33%' }}>
                    <Gauge
                        context="primary"
                        percentage={80}
                        value="$54,321"
                        isProgressLoading={isProgressLoading}
                        isContentLoading={isContentLoading}
                    >
                        {isContentLoading ? '—' : 'High'}
                    </Gauge>
                </div>
            </div>
        );
    });
