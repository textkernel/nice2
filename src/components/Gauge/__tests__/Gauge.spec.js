import React from 'react';
import toJson from 'enzyme-to-json';
import { Gauge } from '../Gauge';

describe('Gauge', () => {
    it('should render correctly', () => {
        const wrapper = shallow(
            <Gauge percentage={33} value="Value" metric="Metric">
                Some content
            </Gauge>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render with loading state correctly', () => {
        const wrapper = shallow(
            <Gauge percentage={33} value="Value" metric="Metric" isContentLoading isProgressLoading>
                Some content
            </Gauge>
        );
        expect(wrapper.find('ContentPlaceholder').exists()).toBeTruthy();
    });
});
