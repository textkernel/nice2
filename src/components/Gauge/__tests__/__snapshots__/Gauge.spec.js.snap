// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gauge should render correctly 1`] = `
<div
  className="Gauge"
>
  <svg
    className="Gauge__svg"
    width="100%"
  >
    <linearGradient
      id="Gauge__gradient--brand"
    >
      <stop
        className="Gauge__gradientStart--brand"
        offset="50%"
      />
      <stop
        className="Gauge__gradientEnd--brand"
        offset="100%"
      />
    </linearGradient>
    <circle
      className="Gauge__circleBackground"
      cx="50%"
      cy="25%"
      r={100}
      style={
        Object {
          "strokeDasharray": "314.1592653589793 628.3185307179587",
        }
      }
    />
    <circle
      className="Gauge__circleForeground"
      cx="50%"
      cy="25%"
      r={100}
      stroke="url(#Gauge__gradient--brand)"
      style={
        Object {
          "strokeDasharray": "103.67255756846318 628.3185307179587",
        }
      }
    />
  </svg>
  <div
    className="Gauge__contentWrapper"
  >
    <span
      className="Gauge__value"
    >
      Value
      <span
        className="Gauge__metric"
      >
        Metric
      </span>
    </span>
  </div>
  <div
    className="Gauge__bottom"
  >
    Some content
  </div>
</div>
`;
